src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -1030,18 +1035,14 @@ private static CachedCodeEntry LookupCachedAndUpdate(CachedCodeEntryKey key)
         {
             lock (s_livecode)
             {
-                for (LinkedListNode<CachedCodeEntry> current = s_livecode.First; current != null; current = current.Next)
+                if (s_livecode_dict.ContainsKey(key))
                 {
-                    if (current.Value._key == key)
-                    {
-                        // If we find an entry in the cache, move it to the head at the same time.
-                        s_livecode.Remove(current);
-                        s_livecode.AddFirst(current);
-                        return current.Value;
-                    }
+                    var entry = s_livecode_dict[key];
+                    s_livecode.Remove(entry);
+                    s_livecode.AddFirst(entry);
```

Comment:
Each time we do this Remove/AddFirst, we end up allocating another node in the linked list.  That's one of the reasons I suggest just making the entries themselves into nodes in their own list, and avoiding `LinkedList<T>`.